File: bin/zircon_benchmarks/null.cc
// Copyright 2017 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "test_runner.h"

namespace {

// This is a test that does nothing.  This is useful for measuring the
// overhead of the performance testing framework.  There will be some
// overhead involved in calling the function (via a direct or indirect
// call if the call is not inlined) and collecting timing data around
// the function call.
void NullTest() {}

__attribute__((constructor)) void RegisterTests() {
  fbenchmark::RegisterTestFunc<NullTest>("Null");
}

}  // namespace

File: src/proc/bin/starnix/device/null.rs
// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use crate::fd_impl_seekable;
use crate::fs::*;
use crate::task::*;
use crate::types::*;

pub struct DevNull;

impl DevNull {
    pub const MINOR: u32 = 3;
}

impl FileOps for DevNull {
    fd_impl_seekable!();

    fn write_at(
        &self,
        _file: &FileObject,
        _task: &Task,
        _offset: usize,
        data: &[UserBuffer],
    ) -> Result<usize, Errno> {
        Ok(data.len())
    }

    fn read_at(
        &self,
        _file: &FileObject,
        _task: &Task,
        _offset: usize,
        _data: &[UserBuffer],
    ) -> Result<usize, Errno> {
        Ok(0)
    }
}
